src/device-state/current-state.ts

```diff
@@ -130,100 +140,124 @@ const getStateDiff = (): DeviceStatus => {
 	return _.omitBy(diff, _.isEmpty);
 };
 
-const report = _.throttle(async () => {
-	const conf = await config.getMany([
-		'deviceApiKey',
-		'apiTimeout',
-		'apiEndpoint',
-		'uuid',
-		'localMode',
-	]);
-
-	const stateDiff = getStateDiff();
-	if (_.size(stateDiff) === 0) {
-		return 0;
-	}
+const throttledReport = _.throttle(
+	// We define the throttled function this way to avoid UncaughtPromise exceptions
+	// for exceptions thrown from the report function
+	(opts: StateReport, resolve: () => void, reject: (e: Error) => void) =>
+		report(opts).then(resolve).catch(reject),
+	constants.maxReportFrequency,
+);
 
-	if (conf.uuid == null || conf.apiEndpoint == null) {
-		throw new InternalInconsistencyError(
-			'No uuid or apiEndpoint provided to CurrentState.report',
-		);
-	}
-
-	try {
-		await sendReportPatch(stateDiff, conf);
-
-		stateReportErrors = 0;
-		_.assign(lastReportedState.local, stateDiff.local);
-		_.assign(lastReportedState.dependent, stateDiff.dependent);
-	} catch (e) {
-		if (e instanceof StatusError) {
-			// We don't want this to be classed as a report error, as this will cause
-			// the watchdog to kill the supervisor - and killing the supervisor will
-			// not help in this situation
-			log.error(
-				`Non-200 response from the API! Status code: ${e.statusCode} - message:`,
-				e,
-			);
-		} else {
-			throw e;
-		}
-	}
-}, constants.maxReportFrequency);
+/**
+ * Perform exponential backoff on the function, increasing the attempts
+ * counter on each call
+ *
+ * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
+ */
+const backoff = (
+	retry: (attempts: number) => void,
+	attempts = 0,
+	maxDelay: number,
+	minDelay = MINIMUM_BACKOFF_DELAY,
+) => {
+	const delay = Math.min(2 ** attempts * minDelay, maxDelay);
+	log.info(`Retrying request in ${delay / 1000} seconds`);
+	setTimeout(() => retry(attempts + 1), delay);
+};
```

Comment:
Is there a way to reset this backoff delay in current state reporting, such as with an API call that uses `force`?

Reply:
I'm imagining a scenario where current state reporting has failed and is at the maximum backoff delay time. The user then changed something in their network, since perhaps that was the issue that caused state reporting to fail, and would like to force a retry of the current state report. How would they achieve this?

Reply:
That's understandable, and since this is the case, it sounds fine to not allow a force option. This LGTM